File: python/61_旋转链表_medium.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return head
        dummyHead = ListNode(0, head)
        cur = dummyHead
        num = 0
        while cur.next:
            num += 1
            cur = cur.next
        if not k % num:
            return dummyHead.next
        k %= num
        i = 0
        cur.next = dummyHead.next
        cur = dummyHead
        while cur.next and i < num-k:
            i += 1
            cur = cur.next
        res = ListNode(0, cur.next)
        cur.next = None
        return res.next

File: cpp/61_旋转链表_medium.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head) return head;
        ListNode *dummyHead = new ListNode(0, head);
        ListNode *cur = dummyHead;
        int num = 0;
        while (cur->next) {
            num++;
            cur = cur->next;
        }
        if (k % num == 0) return dummyHead->next;
        k %= num;
        cur->next = dummyHead->next;
        // ListNode *res = new ListNode(0, cur);
        cur = dummyHead;
        int i = 0;
        while (cur->next && i<num-k) {
            cur = cur->next;
            i++;
        }
        ListNode *p = cur->next;
        cur->next = nullptr;
        return p;
    }
};